### PR TITLE
Don't simplify Nullable{T} to T? inside crefs

### LINKED ISSUE #29: Incorrect simplification of Nullable{T} in XML comment
The following shows an incorrect **Simplify name** proposal for a reference to `Nullable<T>` within an XML documentation comment.

image

This can be reproduced in Visual Studio 2015 Preview (CTP 4?) and CTP 5 with the following code:

```csharp
/// <summary>
/// The <see cref="Nullable{T}"/> type has been defined not using the C# shorthand.
/// </summary>
public class Foo
{
}
```

### PR DESCRIPTION
**Pull request by shyamnamboodiripad**
Fixes #29

### CODE REVIEW

**Comment by shyamnamboodiripad** [comment]
@srivatsn @heejaechang @mavasani @jmarolf @tmeschter  Could you please take a look?

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 1]
```diff
@@ -1340,7 +1340,8 @@ private static int GetNamespaceId(SyntaxList<MemberDeclarationSyntax> members, N
 
                     // nullable rewrite: Nullable<int> -> int?
                     // Don't rewrite in the case where Nullable<int> is part of some qualified name like Nullable<int>.Something
-                    if (!name.IsVar && (symbol.Kind == SymbolKind.NamedType) && !name.IsLeftSideOfQualifiedName())
+                    // Don't rewrite cref="Nullable{T}" to cref="T?" since the ? syntax is not supported within crefs
```
Comment (sharwell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: This comment is misleading. The `?` syntax is supported in crefs, as long as you are referencing a constructed generic type. For example, the following works fine:

```xml
<see cref="Enumerable.Sum(IEnumerable{decimal?})"/>
```

The problem with `<see cref="Nullable{T}"/>` is you are actually referencing an open generic type. This is exactly the same situation as not being able to use the shorthand syntax for `typeof(Nullable<>)` (which also references the open generic type).

File: src/Workspaces/VisualBasic/Portable/Extensions/ExpressionSyntaxExtensions.vb [hunk 2]
```diff
@@ -1291,7 +1291,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
 
                     ' Nullable rewrite: Nullable(Of Integer) -> Integer?
                     ' Don't rewrite in the case where Nullable(Of Integer) is part of some qualified name like Nullable(Of Integer).Something
-                    If (symbol.Kind = SymbolKind.NamedType) AndAlso (Not name.IsLeftSideOfQualifiedName) Then
+                    ' Don't rewrite cref="Nullable(Of T)" to cref="T?" since the ? syntax is not supported within crefs
```
Comment (sharwell) [inline src/Workspaces/VisualBasic/Portable/Extensions/ExpressionSyntaxExtensions.vb]: :bulb: See my note on the previous comment.

**Comment by jmarolf** [comment]
:+1:

**Comment by sharwell** [comment]
:exclamation: This change will cause the editor to stop trying to simplify this:

```xml
<see cref="Enumerable.Sum(IEnumerable{Nullable<decimal>})"/>
```

To this:

```xml
<see cref="Enumerable.Sum(IEnumerable{decimal?})"/>
```

We *do* want to support this simplification. Rather than simply use `InsideCrefReference`, you should use the set of conditions [shown here](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/a30e6f6adddebe24bd5100029ee551d9e35ff2c9/StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1125UseShorthandForNullableTypes.cs#L66-L81) to identify cases where `Nullable{T}` can't be simplified.

**Comment by shyamnamboodiripad** [comment]
@sharwell thanks much for your comments and pointers above. In my original testing of this issue, I incorrectly assumed that the ```?``` syntax is not supported inside ```crefs``` at all - but I understand the issue better now. I will look into how you solved this for the style cop analyzers above and update this PR accordingly. Thanks! :+1:

**Comment by shyamnamboodiripad** [comment]
@sharwell looking at the code you linked above, it seems like the checks you have are purely syntactic. In the case of the simplifier, we have access to semantic information which would allow us to check whether or not the nullable type we are trying to simplify is a constructed generic type and offer the simplification only if it is.

In other words, we can check the ```ITypeSymbol.DeclaringSyntaxReferences``` of ```XYZ``` in ```<see cref="Nullable{XYZ}">``` to see whether ```XYZ``` is being declared inside the same ```cref```. If yes, we are in an unconstructed generic type and we won't offer the simplification. If not (e.g., ```<see cref="SomeType.SomeMethod(Nullable{XYZ})"/>```) we will offer the simplification.

Do you see any issues with this approach?

Also including @amcasey to see if he has an opinion.

**Comment by sharwell** [comment]
@shyamnamboodiripad Last I checked, the IDE (incorrectly) reports that `<see cref="Nullable{Int32}"/>` is referencing a generic instance. You'll want to make sure this is not the case.

**Comment by shyamnamboodiripad** [comment]
Updated commit addresses issues raised above. I ran into some issues on compiler side while trying this fix this - I've logged #2170, #2189, #2196 and #2197 for these issues.

@srivatsn @tmeschter @mavasani @heejaechang could you please take a look?

**Comment by tmeschter** [comment]
:+1:

**Comment by sharwell** [comment]
@shyamnamboodiripad Sounds like we make the right decision going with the syntax approach. :smile:  But I'm glad to see these fixed.

**Comment by shyamnamboodiripad** [comment]
😊 Yes. In C# it is at least possible to use a semantics based approach. But VB currently needs to check syntax because of the above bugs.

I think it should always be possible to use semantics for type checking in any context in the languages and SemanticModel should ideally never disagree with compiler like in the above bugs. So in some ways, although that ended up delaying this fix, its good that we found these bugs 😊

**Comment by shyamnamboodiripad** [comment]
@srivatsn @heejaechang - I need one more thumbs up to merge this PR. Could you please take a quick look?

**Comment by srivatsn** [comment]
:+1:

**Pull request merged**